### PR TITLE
Sidebar - change minWidth prop, add defaultWidth, add resize callback 

### PR DESCRIPTION
**Pull request by SergeyBondar93**

### CODE REVIEW

File: packages/sidebar/src/interfaces.ts [hunk 1]
```diff
@@ -28,6 +28,12 @@ export interface ISidebarProps {
   maxWidth?: number;
   navWidth?: number;
   separatorWidth?: number;
+  rollWidth?: number;
+  onChangeWidth?: Function;
+  params?: {
```
Comment (tundraray) [inline packages/sidebar/src/interfaces.ts]: Props Params not needed. Collapsed and width must be on the same level as other props

File: packages/sidebar/src/interfaces.ts [hunk 2]
```diff
@@ -28,6 +28,12 @@ export interface ISidebarProps {
   maxWidth?: number;
   navWidth?: number;
   separatorWidth?: number;
+  rollWidth?: number;
```
Comment (tundraray) [inline packages/sidebar/src/interfaces.ts]: maybe. it is better to call it a minimizedWidth

File: packages/sidebar/stories/Sidebar.stories.tsx [hunk 3]
```diff
@@ -287,4 +290,40 @@ storiesOf('Sidebar', module)
...
+        ? JSON.parse(sSParams)
+        : {
+          collapsed: false,
+          width: 300,
+        },
+    );
+
+
+    const onChangeWidth = (newParams) => {
+      console.log(newParams);
```
Comment (tundraray) [inline packages/sidebar/stories/Sidebar.stories.tsx]: console.log

File: packages/sidebar/src/Sidebar.tsx [hunk 4]
```diff
@@ -36,67 +36,85 @@ export const PureSidebar: React.FC<ISidebarProps> = ({
...
     e.preventDefault();
   }, []);
 
-  const handleMouseUp = useCallback(() => {
+  const handleMouseUp = useCallback((e) => {
+    if (widthRef.current <= minWidth) {
+      setWidth(rollWidth);
+      setAnimate(true);
+    }
+    if (!e.target.closest('button')) {
```
Comment (ISp1rit) [inline packages/sidebar/src/Sidebar.tsx]: This is very strange logic. What is it for?

File: packages/sidebar/src/Sidebar.tsx [hunk 5]
```diff
@@ -36,67 +36,85 @@ export const PureSidebar: React.FC<ISidebarProps> = ({
...
+        collapsed: widthRef.current < minWidth,
+        width: widthRef.current < minWidth ? minWidth + 1 : widthRef.current,
+      });
+    }
     document.removeEventListener('selectstart', handleSelectStart);
     document.removeEventListener('mouseup', handleMouseUp);
     document.removeEventListener('mousemove', handleMouseMove);
     setAntiSelectLayer(false);
-  }, [handleSelectStart, handleMouseMove]);
+  }, [handleSelectStart, handleMouseMove, minWidth, widthRef.current, rollWidth, width]);
```
Comment (ISp1rit) [inline packages/sidebar/src/Sidebar.tsx]: Any ref is unnecessary dependency. so widthRef.current - mustn't be here.

File: packages/sidebar/src/Sidebar.tsx [hunk 6]
```diff
@@ -112,15 +130,16 @@ export const PureSidebar: React.FC<ISidebarProps> = ({
...
+      setWidth(propsWidth);
     } else {
-      setWidth(minWidth);
-      setArrowToRight(true);
+      setWidth(rollWidth);
     }
-  }, [arrowToRight, maxWidth, minWidth]);
+    onChangeWidth({
+      collapsed: !arrowToRight,
+      width: widthRef.current < minWidth ? minWidth + 1 : widthRef.current,
```
Comment (ISp1rit) [inline packages/sidebar/src/Sidebar.tsx]: Logically, there must be width: arrowToRight ? propsWidth : rollWidth.

File: packages/sidebar/src/Sidebar.tsx [hunk 7]
```diff
@@ -36,67 +36,85 @@ export const PureSidebar: React.FC<ISidebarProps> = ({
...
+  }, [collapsed, propsWidth]);
 
   const handleMouseMove = useCallback(
     (e) => {
       const { clientX: currentX } = e;
 
       const newWidth = isRTL
-        ? width - (currentX - clickXRef.current)
-        : width + (currentX - clickXRef.current);
+        ? Number(width) - Number(currentX - clickXRef.current)
```
Comment (ISp1rit) [inline packages/sidebar/src/Sidebar.tsx]: Why you add Number? This values already numbers.

File: packages/sidebar/stories/Sidebar.stories.tsx (hunk 3 above)
Comment (SergeyBondar93) [reply packages/sidebar/stories/Sidebar.stories.tsx]: all right, this is storybook

File: packages/sidebar/src/Sidebar.tsx (hunk 7 above)
Comment (tundraray) [reply packages/sidebar/src/Sidebar.tsx]: yep. just add type **MouseEvent**  for **e** variable

File: packages/sidebar/src/Sidebar.tsx [hunk 8]
```diff
@@ -36,67 +36,85 @@ export const PureSidebar: React.FC<ISidebarProps> = ({
...
 
   const sidebarRef = useRef<HTMLDivElement>(null);
   const observerRef: React.MutableRefObject<ResizeObserver | undefined> = useRef();
   const clickXRef = useRef(0);
-  const widthRef = useRef(maxWidth);
+  const widthRef = useRef(propsWidth);
+
+  useEffect(() => {
+    setArrowToRight(width < minWidth);
+  }, [minWidth, propsWidth, width]);
```
Comment (tundraray) [inline packages/sidebar/src/Sidebar.tsx]: Is it really necessary to add variables for dependencies here?

File: packages/sidebar/src/Sidebar.tsx (hunk 6 above)
Comment (SergeyBondar93) [reply packages/sidebar/src/Sidebar.tsx]: only  `width: propsWidth,` :)

File: packages/sidebar/src/Sidebar.tsx [hunk 9]
```diff
@@ -36,67 +36,87 @@ export const PureSidebar: React.FC<ISidebarProps> = ({
...
   const observerRef: React.MutableRefObject<ResizeObserver | undefined> = useRef();
   const clickXRef = useRef(0);
-  const widthRef = useRef(maxWidth);
+  const widthRef = useRef(propsWidth);
+
+
+  useEffect(() => {
+    setWidth(collapsed ? minimizedWidth : propsWidth);
+    setArrowToRight(collapsed);
+  }, [collapsed, propsWidth]);
```
Comment (ISp1rit) [inline packages/sidebar/src/Sidebar.tsx]: There must be minimizedWidth too.

**Pull request merged**